src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGenerator.cs

```diff
@@ -881,12 +884,51 @@ private OpenApiSchema GenerateSchemaFromFormParameters(
                 Required = new SortedSet<string>(requiredPropertyNames)
             };
         }
+        private static IList<ApiResponseType> GetResponseTypes(ApiDescription apiDescription)
+        {
+#if NET7_0_OR_GREATER
+            var supportedResponseTypes = new List<ApiResponseType>(apiDescription.SupportedResponseTypes);
+            if (apiDescription.ActionDescriptor is ControllerActionDescriptor controllerActionDescriptor)
+            {
+                var returnType = UnwrapTask(controllerActionDescriptor.MethodInfo.ReturnType);
+                if (typeof(IEndpointMetadataProvider).IsAssignableFrom(returnType))
+                {
+                    var populateMetadataMethod = returnType.GetMethod("Microsoft.AspNetCore.Http.Metadata.IEndpointMetadataProvider.PopulateMetadata", BindingFlags.Static | BindingFlags.NonPublic);
+                    if (populateMetadataMethod != null)
+                    {
+                        var endpointBuilder = new MetadataEndpointBuilder();
+                        populateMetadataMethod.Invoke(null, [controllerActionDescriptor.MethodInfo, endpointBuilder]);
+                        var responseTypes = endpointBuilder.Metadata.Cast<IProducesResponseTypeMetadata>().ToList();
+                        foreach (var responseType in responseTypes)
+                        {
+                            supportedResponseTypes.Add(
+                                new ApiResponseType()
+                                {
+                                    IsDefaultResponse = false,
+                                    Type = responseType.Type,
+                                    StatusCode = responseType.StatusCode,
+                                    ApiResponseFormats = responseType.ContentTypes.Select(contentType => new ApiResponseFormat { MediaType = contentType }).ToList()
+                                });
+                        }
+                    }
+                }
+            }
+
+            static Type UnwrapTask(Type type)
+            => type.IsGenericType && type.GetGenericTypeDefinition() == typeof(Task<>)
+                ? type.GetGenericArguments()[0]
+                : type;
+            return supportedResponseTypes;
```

Comment:
IMHO it's more readable to not embed static methods in the middle of the code using them.

```suggestion
            return supportedResponseTypes;

            static Type UnwrapTask(Type type)
                => type.IsGenericType && type.GetGenericTypeDefinition() == typeof(Task<>)
                    ? type.GetGenericArguments()[0]
                    : type;
```